### PR TITLE
Project join and cancel join request

### PR DESCRIPTION
**Pull request by subinasr**
- Addresses #1980 and #1981

## Changes

* Handle Project join and cancel join request in Explore DEEP page

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [ ] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

console.warn has been used to notify success and failures of requests in absence of notify

### CODE REVIEW

File: app/views/ExploreDeep/ProjectFilterForm/styles.css [hunk 1]
```diff
@@ -0,0 +1,19 @@
+.content {
+    display: flex;
+
+    .filters {
+        display: flex;
+        align-items: center;
+        flex-grow: 1;
+        flex-wrap: wrap;
+        padding: var(--dui-spacing-medium);
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/styles.css]: Do we need to add padding?

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    const handleClearFilters = useCallback(() => {
+        onFiltersChange({});
+    }, [onFiltersChange]);
+
+    /*
+    const [
+        organizationOptions,
+        setOrganizationOptions,
+    ] = useState<OrganizationDetails | undefined>();
+    */
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx]: Is this a wip?

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    return (
+        <div className={_cs(styles.actionCell, className)}>
+            {!membershipPending ? (
+                <Button
+                    name="join"
+                    onClick={showJoinModal}
+                    variant="secondary"
+                    disabled={disabled}
+                    title="Join Project"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]: We don't need title if the button already has a label

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                >
+                    Join
+                </Button>
+            ) : (
+                <ConfirmButton
+                    name="cancel-join"
+                    onConfirm={handleCancelJoinProjectClick}
+                    variant="secondary"
+                    disabled={disabled}
+                    title="Cancel Join Request"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]: We don't need title if the button already has a label

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                    onClick={showJoinModal}
+                    variant="secondary"
+                    disabled={disabled}
+                    title="Join Project"
+                >
+                    Join
+                </Button>
+            ) : (
+                <ConfirmButton
+                    name="cancel-join"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]:
```suggestion
                    name={undefined}
```

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    if (isMember) {
+        return null;
+    }
+
+    return (
+        <div className={_cs(styles.actionCell, className)}>
+            {!membershipPending ? (
+                <Button
+                    name="join"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]:
```suggestion
                    name={undefined}
```

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        cancelJoinProject,
+    ] = useMutation<CancelJoinProjectMutation, CancelJoinProjectMutationVariables>(
+        CANCEL_JOIN_PROJECT,
+        {
+            onCompleted: () => {
+                console.warn('Join Request successfully sent');
+            },
+            onError: () => {
+                console.warn('Failed to send request');
+            },
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]: Use notify

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    );
+
+    return (
+        <Modal
+            heading="Send project join request"
+            className={_cs(className, styles.projectJoinModal)}
+            onCloseButtonClick={onModalClose}
+            footerActions={(
+                <>
+                    <Button
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx:159]: Add NonFieldError and PendingAnimation

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        () => {
+            const { errored, error: err, value: val } = validate();
+            setError(err);
+            if (!errored && isDefined(val)) {
+                joinProject({
+                    variables: valueToSend,
+                });
+            }
+        },
+        [setError, validate, projectId, value],
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx]: Why add projectId as dependency here?

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    ] = useMutation<JoinProjectMutation, JoinProjectMutationVariables>(
+        JOIN_PROJECT,
+        {
+            onCompleted: () => {
+                console.warn('Join Request successfully sent');
+                onModalClose();
+            },
+            onError: () => {
+                console.warn('Failed to send request');
+            },
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx]: Use notify

Comment (tnagorra) [reply app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx]: Also handle error from server

File: app/components/FrameworkImageButton/index.tsx [hunk 11]
```diff
@@ -33,6 +33,7 @@ function FrameworkImageButton(props: Props) {
         label,
         variant,
     } = props;
+    console.warn('here', props);
```
Comment (tnagorra) [inline app/components/FrameworkImageButton/index.tsx]: No console.warn meant for debugging

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx (hunk 8 above)
Comment (AdityaKhatri) [reply app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx:159]: This is done

File: app/components/framework/FrameworkImageButton/index.tsx [hunk 12]
```diff
@@ -19,8 +19,8 @@ const query = {
     fields: ['title', 'preview_image', 'id'],
 };
 
-interface Props {
-    frameworkId: number | string;
+export interface Props {
+    frameworkId?: number | string;
```
Comment (tnagorra) [inline app/components/framework/FrameworkImageButton/index.tsx]: Why change this typing? It's always expected to pass a frameworkId for FrameworkImageButton

Comment (tnagorra) [reply app/components/framework/FrameworkImageButton/index.tsx]: Better to use `frameworkId: number | string | undefined`

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                    placeholder="any"
+                    value={value?.analysisFrameworks}
+                    onChange={setFieldValue}
+                    options={analysisFrameworkOptions}
+                    onOptionsChange={setAnalysisFrameworkOptions}
+                />
+            </div>
+            <div className={styles.buttonContainer}>
+                <Button
+                    name="submit"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx]:
```suggestion
                    name={undefined}
```

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+                    label="Organizations"
+                    value={value?.organizations}
+                    onChange={setFieldValue}
+                    options={organizationOptions}
+                    onOptionsChange={setOrganizationOptions}
+                    placeholder="any"
+                />
+                <FrameworkMultiSelectInput
+                    name="analysisFrameworks"
+                    label="Analysis Framework"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx]:
```suggestion
                    label="Analysis Frameworks"
```

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+    return (
+        <div className={styles.content}>
+            <div className={styles.filters}>
+                <TextInput
+                    name="search"
+                    label="Search"
+                    value={value?.search}
+                    onChange={setFieldValue}
+                    placeholder="any"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx:97]:
```suggestion
```

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+                    onChange={setFieldValue}
+                />
+                <OrganizationMultiSelectInput
+                    name="organizations"
+                    label="Organizations"
+                    value={value?.organizations}
+                    onChange={setFieldValue}
+                    options={organizationOptions}
+                    onOptionsChange={setOrganizationOptions}
+                    placeholder="any"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx:118]:
```suggestion
```

File: app/views/ExploreDeep/ProjectFilterForm/index.tsx [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+                />
+                <FrameworkMultiSelectInput
+                    name="analysisFrameworks"
+                    label="Analysis Framework"
+                    placeholder="any"
```
Comment (tnagorra) [inline app/views/ExploreDeep/ProjectFilterForm/index.tsx:123]:
```suggestion
```

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+import ProjectJoinModal from './ProjectJoinModal';
+
+import styles from './styles.css';
+
+export interface Props {
+    className?: string;
+    projectId: string;
+    disabled?: boolean;
+    membershipPending: boolean;
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx:26]: Let's add a note to indicate that this is not actual data pending for membership

File: app/views/ExploreDeep/ActionCell/index.tsx [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+                onMemberStatusChange();
+                alert.show(
+                    'Successfully sent join request.',
+                    {
+                        variant: 'success',
+                    },
+                );
+            },
+        },
+    );
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/index.tsx]: Let's handle errors

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+    const error = getErrorObject(riskyError);
+
+    const [
+        joinProject,
+        { loading },
+    ] = useMutation<JoinProjectMutation, JoinProjectMutationVariables>(
+        JOIN_PROJECT,
+        {
+            onCompleted: (response) => {
```
Comment (tnagorra) [inline app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx:100]: We need to handle onError condition as well

File: app/components/selections/NewOrganizationMultiSelectInput/index.tsx [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+type OrganizationMultiSelectInputProps<K extends string> = SearchMultiSelectInputProps<
+    string,
+    K,
+    BasicOrganization,
+    Def,
+    'onSearchValueChange' | 'searchOptions' | 'optionsPending' | 'keySelector' | 'labelSelector' | 'totalOptionsCount' | 'onShowDropdownChange'
+>;
+
+const keySelector = (d: BasicOrganization) => d.id;
+const labelSelector = (d: BasicOrganization) => d.title;
```
Comment (tnagorra) [inline app/components/selections/NewOrganizationMultiSelectInput/index.tsx]: I am always confused about this. Should the title be `title` or `mergeAs.title`?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/components/selections/NewOrganizationMultiSelectInput/index.tsx (hunk 21 above)
Comment (AdityaKhatri) [reply app/components/selections/NewOrganizationMultiSelectInput/index.tsx]: This is fine, as new searches won't return merged values

File: app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx (hunk 20 above)
Comment (AdityaKhatri) [reply app/views/ExploreDeep/ActionCell/ProjectJoinModal/index.tsx:100]: There should be a generic onFailure imo

File: app/components/selections/NewOrganizationMultiSelectInput/index.tsx (hunk 21 above)
Comment (tnagorra) [reply app/components/selections/NewOrganizationMultiSelectInput/index.tsx]: Noh!

**Review by tnagorra — DISMISSED** [review]

**Review by tnagorra — APPROVED** [review]

**Pull request merged**